Parser rejects most LLM-emitted observation types → everything mis-tagged as "bugfix"; no retention/prune path for unbounded DB growth

Two issues observed on v13.1.0 (macOS, `code` mode, ~2 months of use):

## 1. Observation-type drift → silent mis-tagging

The summarizer LLM routinely emits `type` values outside the active mode's
`observation_types` enum. The parser falls back to `observation_types[0]` — which in
`code` mode is `bugfix` — so a large share of observations are stored mis-tagged.

One day's log (`claude-mem-2026-07-07.log`) shows **445** `[ERROR] [PARSER] Invalid
observation type` lines. Frequency tally over recent logs:
`code` 63, `milestone` 39, `test` 34, `state` 27, `fix` 21, `what` 20, `architecture` 20,
`bug` 18, `task` 15, `finding` 14, `tool` 10, `skeptic` 10, plus a long tail.

Impact: type-filtered retrieval is degraded — most of the store reads as "bugfix".

Suggestions (either helps):
- Fall back to a neutral type (`change`) instead of `observation_types[0]`, or make the
  fallback configurable.
- Strengthen the summarizer prompt's type constraint, and/or fuzzy-map near-misses
  (`bug`/`fix` → `bugfix`, `architecture` → closest) before falling back.

## 2. No retention/prune path — unbounded growth

`~/.claude-mem/` reached **3.6 GB in ~2 months** (claude-mem.db 1.4 GB + chroma 2.1 GB).
I could find no TTL/retention/VACUUM logic in the bundled scripts, and `worker-cli.js`
exposes only `start|stop|restart|status`. The only shrink on record was the one-time
v12.4.3 migration wipe. The context-injection knobs (`CLAUDE_MEM_CONTEXT_*`) limit
read-back, not storage.

Suggestion: a `worker-cli.js prune --older-than <days>` (SQLite delete + VACUUM + Chroma
collection cleanup), or a documented supported manual procedure.

Happy to provide logs/DB stats. Thanks for the plugin — the multi-session `.in_use`
PID+procStart locking in 13.x has been working well.

--- a/src/sdk/parser.ts
+++ b/src/sdk/parser.ts
@@ -107,10 +107,9 @@ function parseObservationBlocks(text: string, correlationId?: string | number): 
     const fallbackType = validTypes[0];
     let finalType = fallbackType;
     if (type) {
-      if (validTypes.includes(type.trim())) {
-        finalType = type.trim();
-      } else {
-        logger.error('PARSER', `Invalid observation type: ${type}, using "${fallbackType}"`, { correlationId });
+      finalType = type;
+      if (!validTypes.includes(type)) {
+        logger.error('PARSER', `Invalid observation type: ${type}, preserving emitted type`, { correlationId });
       }
     } else {
       logger.error('PARSER', `Observation missing type field, using "${fallbackType}"`, { correlationId });

--- a/tests/integration/worker-api-endpoints.test.ts
+++ b/tests/integration/worker-api-endpoints.test.ts
@@ -24,6 +24,7 @@ import {
 } from '../../src/shared/dependency-health.js';
 
 let loggerSpies: ReturnType<typeof spyOn>[] = [];
+const serialIt = it.serial;
 
 describe('Worker API Endpoints Integration', () => {
   let server: Server;
@@ -75,7 +76,7 @@ describe('Worker API Endpoints Integration', () => {
 
   describe('Health/Readiness/Version Endpoints', () => {
     describe('GET /api/health', () => {
-      it('should return status, initialized, mcpReady, platform, pid', async () => {
+      serialIt('should return status, initialized, mcpReady, platform, pid', async () => {
         server = new Server(mockOptions);
         await server.listen(testPort, '127.0.0.1');
 
@@ -92,7 +93,7 @@ describe('Worker API Endpoints Integration', () => {
         expect(typeof body.pid).toBe('number');
       });
 
-      it('should reflect uninitialized state', async () => {
+      serialIt('should reflect uninitialized state', async () => {
         const uninitOptions: ServerOptions = {
           getInitializationComplete: () => false,
           getMcpReady: () => false,
@@ -113,7 +114,7 @@ describe('Worker API Endpoints Integration', () => {
         expect(body.mcpReady).toBe(false);
       });
 
-      it('includes dependency health and stays HTTP 200 for dependency-only degradation', async () => {
+      serialIt('includes dependency health and stays HTTP 200 for dependency-only degradation', async () => {
         recordDependencyStatus(
           'uvx',
           'vector_search_unavailable',
@@ -144,7 +145,7 @@ describe('Worker API Endpoints Integration', () => {
     });
 
     describe('GET /api/readiness', () => {
-      it('should return 200 with status ready when initialized', async () => {
+      serialIt('should return 200 with status ready when initialized', async () => {
         server = new Server(mockOptions);
         await server.listen(testPort, '127.0.0.1');
 
@@ -156,7 +157,7 @@ describe('Worker API Endpoints Integration', () => {
         expect(body.mcpReady).toBe(true);
       });
 
-      it('should return 503 with status initializing when not ready', async () => {
+      serialIt('should return 503 with status initializing when not ready', async () => {
         const uninitOptions: ServerOptions = {
           getInitializationComplete: () => false,
           getMcpReady: () => false,
@@ -179,7 +180,7 @@ describe('Worker API Endpoints Integration', () => {
     });
 
     describe('GET /api/version', () => {
-      it('should return version string', async () => {
+      serialIt('should return version string', async () => {
         server = new Server(mockOptions);
         await server.listen(testPort, '127.0.0.1');
 
@@ -193,7 +194,7 @@ describe('Worker API Endpoints Integration', () => {
     });
 
     describe('GET /api/settings/dependency-health', () => {
-      it('passes through WorkerService initialization guard while initialization is incomplete', async () => {
+      serialIt('passes through WorkerService initialization guard while initialization is incomplete', async () => {
         recordDependencyStatus(
           'claude_cli',
           'setup_required',
@@ -229,7 +230,7 @@ describe('Worker API Endpoints Integration', () => {
   });
 
   describe('Error Handling', () => {
-    it('includes dependency health in admin doctor output', async () => {
+    serialIt('includes dependency health in admin doctor output', async () => {
       recordDependencyStatus(
         'claude_cli',
         'setup_required',
@@ -258,7 +259,7 @@ describe('Worker API Endpoints Integration', () => {
     });
 
     describe('404 Not Found', () => {
-      it('should return 404 for unknown GET routes', async () => {
+      serialIt('should return 404 for unknown GET routes', async () => {
         server = new Server(mockOptions);
         server.finalizeRoutes();
         await server.listen(testPort, '127.0.0.1');
@@ -270,7 +271,7 @@ describe('Worker API Endpoints Integration', () => {
         expect(body.error).toBe('NotFound');
       });
 
-      it('should return 404 for unknown POST routes', async () => {
+      serialIt('should return 404 for unknown POST routes', async () => {
         server = new Server(mockOptions);
         server.finalizeRoutes();
         await server.listen(testPort, '127.0.0.1');
@@ -283,7 +284,7 @@ describe('Worker API Endpoints Integration', () => {
         expect(response.status).toBe(404);
       });
 
-      it('should return 404 for nested unknown routes', async () => {
+      serialIt('should return 404 for nested unknown routes', async () => {
         server = new Server(mockOptions);
         server.finalizeRoutes();
         await server.listen(testPort, '127.0.0.1');
@@ -294,7 +295,7 @@ describe('Worker API Endpoints Integration', () => {
     });
 
     describe('Method handling', () => {
-      it('should handle OPTIONS requests', async () => {
+      serialIt('should handle OPTIONS requests', async () => {
         server = new Server(mockOptions);
         await server.listen(testPort, '127.0.0.1');
 
@@ -307,7 +308,7 @@ describe('Worker API Endpoints Integration', () => {
   });
 
   describe('Content-Type Handling', () => {
-    it('should accept application/json content type', async () => {
+    serialIt('should accept application/json content type', async () => {
       server = new Server(mockOptions);
       server.finalizeRoutes();
       await server.listen(testPort, '127.0.0.1');
@@ -321,7 +322,7 @@ describe('Worker API Endpoints Integration', () => {
       expect(response.status).toBe(404);
     });
 
-    it('should return JSON responses with correct content type', async () => {
+    serialIt('should return JSON responses with correct content type', async () => {
       server = new Server(mockOptions);
       await server.listen(testPort, '127.0.0.1');
 
@@ -333,7 +334,7 @@ describe('Worker API Endpoints Integration', () => {
   });
 
   describe('Server State Management', () => {
-    it('should track initialization state dynamically', async () => {
+    serialIt('should track initialization state dynamically', async () => {
       let initialized = false;
       const dynamicOptions: ServerOptions = {
         getInitializationComplete: () => initialized,
@@ -356,7 +357,7 @@ describe('Worker API Endpoints Integration', () => {
       expect(response.status).toBe(200);
     });
 
-    it('should track MCP ready state dynamically', async () => {
+    serialIt('should track MCP ready state dynamically', async () => {
       let mcpReady = false;
       const dynamicOptions: ServerOptions = {
         getInitializationComplete: () => true,
@@ -383,7 +384,7 @@ describe('Worker API Endpoints Integration', () => {
   });
 
   describe('Server Lifecycle', () => {
-    it('should start listening on specified port', async () => {
+    serialIt('should start listening on specified port', async () => {
       server = new Server(mockOptions);
       await server.listen(testPort, '127.0.0.1');
 
@@ -392,7 +393,7 @@ describe('Worker API Endpoints Integration', () => {
       expect(httpServer!.listening).toBe(true);
     });
 
-    it('should close gracefully', async () => {
+    serialIt('should close gracefully', async () => {
       server = new Server(mockOptions);
       await server.listen(testPort, '127.0.0.1');
 
@@ -411,7 +412,7 @@ describe('Worker API Endpoints Integration', () => {
       }
     });
 
-    it('should handle port conflicts', async () => {
+    serialIt('should handle port conflicts', async () => {
       server = new Server(mockOptions);
       const server2 = new Server(mockOptions);
 
@@ -425,7 +426,7 @@ describe('Worker API Endpoints Integration', () => {
       }
     });
 
-    it('should allow restart on same port after close', async () => {
+    serialIt('should allow restart on same port after close', async () => {
       server = new Server(mockOptions);
       await server.listen(testPort, '127.0.0.1');
 
@@ -451,7 +452,7 @@ describe('Worker API Endpoints Integration', () => {
   });
 
   describe('Route Registration', () => {
-    it('should register route handlers', () => {
+    serialIt('should register route handlers', () => {
       server = new Server(mockOptions);
 
       const setupRoutesMock = mock(() => {});
@@ -465,7 +466,7 @@ describe('Worker API Endpoints Integration', () => {
       expect(setupRoutesMock).toHaveBeenCalledWith(server.app);
     });
 
-    it('should register multiple route handlers', () => {
+    serialIt('should register multiple route handlers', () => {
       server = new Server(mockOptions);
 
       const handler1Mock = mock(() => {});

--- a/tests/sdk/parser.test.ts
+++ b/tests/sdk/parser.test.ts
@@ -59,6 +59,7 @@ describe('parseAgentXml — observations', () => {
 
     expect(result).toHaveLength(1);
     expect(result[0].title).toBeNull();
+    expect(result[0].type).toBe('bugfix');
     expect(result[0].narrative).toBe('Patched the null pointer dereference in session handler.');
   });
 
@@ -143,6 +144,18 @@ describe('parseAgentXml — observations', () => {
 
     expect(result).toHaveLength(1);
     expect(result[0].type).toBe('bugfix');
+  });
+
+  it('preserves a reporter-shaped unsupported observation type', () => {
+    const xml = `<observation>
+      <type>code</type>
+      <title>Reporter-shaped unsupported type</title>
+    </observation>`;
+
+    const result = expectObservation(xml);
+
+    expect(result).toHaveLength(1);
+    expect(result[0].type).toBe('code');
   });
 
   it('returns a fail-fast result when no observation/summary blocks are present', () => {

